File: src/main/java/space/maxkonkin/nasapicbot/model/Nasa.java
package space.maxkonkin.nasapicbot.model;

import com.fasterxml.jackson.annotation.JsonProperty;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import tech.ydb.table.result.ResultSetReader;

import java.time.LocalDate;

@Getter
@Setter
@NoArgsConstructor
public class Nasa {
    @JsonProperty("lang")
    private LangCode langCode;

    @JsonProperty("credit")
    private String credit;

    @JsonProperty("copyright")
    private String copyright;

    @JsonProperty("date")
    private LocalDate date;

    @JsonProperty("explanation")
    private String explanation;

    @JsonProperty("hdurl")
    private String hdUrl;

    @JsonProperty("media_type")
    private String mediaType;

    @JsonProperty("service_version")
    private String serviceVersion;

    @JsonProperty("title")
    private String title;

    @JsonProperty("url")
    private String url;

    public Nasa(LangCode langCode, String credit, String copyright, LocalDate date, String explanation,
                String hdUrl, String mediaType, String serviceVersion, String title, String url) {
        this.langCode = langCode;
        this.credit = credit;
        this.copyright = copyright;
        this.date = date;
        this.explanation = explanation;
        this.hdUrl = hdUrl;
        this.mediaType = mediaType;
        this.serviceVersion = serviceVersion;
        this.title = title;
        this.url = url;
    }

    public static Nasa fromResultSet(ResultSetReader resultSet) {
        var date = resultSet.getColumn("date").getDate();
        var langCode = LangCode.valueOf(resultSet.getColumn("lang").getText().toUpperCase());
        var credit = resultSet.getColumn("credit").getText();
        var copyright = resultSet.getColumn("copyright").getText();
        var explanation = resultSet.getColumn("explanation").getText();
        var title = resultSet.getColumn("title").getText();
        var url = resultSet.getColumn("url").getText();
        var hdUrl = resultSet.getColumn("hd_url").getText();
        var mediaType = resultSet.getColumn("media_type").getText();
        return new Nasa(langCode, credit, copyright, date, explanation, hdUrl, mediaType, "v1", title, url);
    }

    @Override
    public String toString() {
        return "Nasa{" +
                "langCode=" + langCode.getCode() +
                ", credit='" + credit + '\'' +
                ", copyright='" + copyright + '\'' +
                ", date='" + date + '\'' +
                ", explanation='" + explanation + '\'' +
                ", hdUrl='" + hdUrl + '\'' +
                ", mediaType='" + mediaType + '\'' +
                ", serviceVersion='" + serviceVersion + '\'' +
                ", title='" + title + '\'' +
                ", url='" + url + '\'' +
                '}';
    }
}

File: src/main/kotlin/space/maxkonkin/nasapicbot/model/Nasa.kt
package space.maxkonkin.nasapicbot.model

import com.fasterxml.jackson.annotation.JsonProperty
import tech.ydb.table.result.ResultSetReader
import java.time.LocalDate

data class Nasa(
    @JsonProperty("lang")
    val langCode: LangCode?,

    @JsonProperty("credit")
    val credit: String?,

    @JsonProperty("copyright")
    val copyright: String?,

    @JsonProperty("date")
    val date: LocalDate?,

    @JsonProperty("explanation")
    val explanation: String?,

    @JsonProperty("hdurl")
    val hdUrl: String?,

    @JsonProperty("media_type")
    val mediaType: String?,

    @JsonProperty("service_version")
    val serviceVersion: String?,

    @JsonProperty("title")
    val title: String?,

    @JsonProperty("url")
    val url: String?
) {
    companion object {
        fun fromResultSet(resultSet: ResultSetReader): Nasa {
            val date = resultSet.getColumn("date").date
            val langCode = LangCode.valueOf(resultSet.getColumn("lang").text.uppercase())
            val credit = resultSet.getColumn("credit").text
            val copyright = resultSet.getColumn("copyright").text
            val explanation = resultSet.getColumn("explanation").text
            val title = resultSet.getColumn("title").text
            val url = resultSet.getColumn("url").text
            val hdUrl = resultSet.getColumn("hd_url").text
            val mediaType = resultSet.getColumn("media_type").text
            return Nasa(
                langCode, credit, copyright, date, explanation, hdUrl, mediaType, "v1", title, url
            )
        }
    }
}
